docs/tutorials/demo_best_practices.py

```diff
@@ -0,0 +1,281 @@
+    return df
+
+
+sample_json_data = {
+    "hourly": [{"time": "2023-01-01 00:00", "rain": 0}, {"time": "2023-01-01 01:00", "rain": 2}]
+}
+
+# Tests
+df_test = create_dataframe_from_json(sample_json_data)
+assert isinstance(df_test, pd.DataFrame), "Output should be a pandas DataFrame"
+assert list(df_test.columns) == ["time", "rain"], "DataFrame should have time and rain columns"
+assert len(df_test) == 2, "DataFrame should have two rows"
+
+
+# %%
+def convert_time_column_to_datetime(df):
+    df["time"] = pd.to_datetime(df["time"])
+    return df
+
+
+# Tests
+df_test = pd.DataFrame({"time": ["2023-01-01 00:00", "2023-01-01 01:00"], "rain": [0, 2]})
+
+df_test = convert_time_column_to_datetime(df_test)
+assert pd.api.types.is_datetime64_any_dtype(
+    df_test["time"]
+), "Time column should be of datetime type"
+
+
+# %%
+def add_rain_indicators(df):
+    df["is_raining"] = df.rain > 0
+    df["is_raining_in_1_day"] = df.is_raining.shift(24).astype(bool)
+    return df.dropna()
+
+
+# Tests
+df_test = pd.DataFrame(
+    {"time": pd.date_range(start="2023-01-01", periods=48, freq="H"), "rain": [0] * 24 + [2] * 24}
+)
+
+# Test the function
+df_test = add_rain_indicators(df_test)
+assert (
+    "is_raining" in df_test.columns and "is_raining_in_1_day" in df_test.columns
+), "Both indicator columns should be present"
+assert all(
+    df_test.loc[24:, "is_raining"]
+), "All values should be True in 'is_raining' for the second day"
+assert all(
+    df_test.loc[:23, "is_raining_in_1_day"]
+), "All values should be True in 'is_raining_in_1_day' for the first day"
+assert df_test.isna().sum().sum() == 0, "There should be no NaN values in the DataFrame"
+
+# %% [markdown]
+"""
+Encapsulating data processing steps in unit tested functions (which are also made available to the whole team or organisation) saves some work by reusing our code.
+Additionally, the tests increase our confidence in the data quality and serve as additional context to infer the purpose of the function should we or someone else come back at a later time.
+
+We can now apply the functions to process the data.
+"""
+
+# %%
+json_data = load_json_data(outfile)
+df = create_dataframe_from_json(json_data)
+df = convert_time_column_to_datetime(df)
+df = add_rain_indicators(df)
+
+df.head(5)
+
+# %% [markdown]
+"""
+## Use commits as checkpoints and merge branches for atomic changes
+
+Now we can commit the updated data to the transform-raw-data branch in lakeFS repository.
+We write a descriptive commit message.
+"""
+# %%
+with fs.transaction as tx:
+    df.to_csv(f"lakefs://{REPO_NAME}/main/weather_2010.csv")
+    commit = tx.commit(repository=REPO_NAME, branch="main", message="Preprocess 2010 data")
+print(commit)
+
+# %% [markdown]
+"""
+We see that the commit has a unique id, the commit SHA which we can log to identify this particular state of the data.
+
+As the data looks good, we can also merge the branch back into `main`.
```

Comment:
"looks good" to me implies that there is some review happening between creating the commit and merging - we should mention this in the text.